src/Native/Unix/System.Native/pal_console.cpp

```diff
@@ -131,7 +131,9 @@ extern "C" void SystemNative_InitializeConsoleBeforeRead(uint8_t minChars, uint8
             g_preReadTermios = newTermios;
         }
 
-        newTermios.c_iflag &= static_cast<uint32_t>(~(IXON | IXOFF));
+        // On Windows, the Return key generates '\r'. We clear ICRNL to get the same behavior on Unix.
+        // Without this we'd get '\n' which causes newlines when echoed.
+        newTermios.c_iflag &= static_cast<uint32_t>(~(IXON | IXOFF | ICRNL));
```

Comment:
Won't this cause other problems, like code which is expecting `Environment.NewLine` (`"\n") while reading from Console.In actually getting `"\r\n"`?

Reply:
You're right. When reading using Console.In this changes newlines from '\n' (Environment.NewLine) to '\r'.

We could make the change here:
https://github.com/dotnet/corefx/blob/619dbe6503a2de2afa27122cb34716d9ff6906ff/src/System.Console/src/System/ConsolePal.Unix.cs#L91-L104
If we see ConsoleKeyInfo for the return key on Unix, we replace it with the Windows values.

Reply:
How does that fix the referenced issue?

Reply:
This is where the char gets printed. It will print \r like on Windows.

Reply:
I'm not understanding.  The repro is doing:
```C#
while (true)
{
    Console.WriteLine(DateTime.Now);
    // do some work
    Console.ReadKey();
}
```
How does the KeyChar value stored in the returned ConsoleKeyInfo affect that?  The returned ConsoleKeyInfo is ignored.

Reply:
ReadKey does Console.Write of KeyChar (line 102 in above snippit).

Reply:
Ah, I see. Hmm.

Reply:
The thing that concerns me here is we're entirely making this up; there is no `\r` anywhere, but we're manufacturing one.  And it's a visible change in that a parent process redirecting the stdout of a child that calls Console.ReadKey could see the child write '\r' instead of '\n'.  Isn't that strange?

Reply:
It seems less strange when doing this by changing the terminal settings than doing it ourselves, but the code comes down to the same thing: replace '\n' with '\r'.

While possible, the parent child setup would be very special: standard input isn't redirected (otherwise ReadKey throws), standard output is, and the child is echoing standard input.

ReadKey is meant for interactive applications (reading&echoing from terminal) and the newline echo is breaking that use-case.

Reply:
It's the echoing you're concerned about?  Would it make more sense then to just make the change be:
```C#
if (!intercept && !previouslyProcessed)
{
    char c = keyInfo.KeyChar;
    if (c == '\n' && keyInfo.Key == ConsoleKey.Enter)
    {
        // Replace the '\n' char for Enter by '\r' to match Windows behavior.
        c = '\r';
    }
    Console.Write(c)
}
```
?

Reply:
I considered that too. It limits the scope even more. I chose to echo and return the same char for consistency (e.g. the user may want to intercept and echo himself).

Reply:
I find it safe to return the '\r' to the user, since that is exactly how it works on Windows. I think some users may expect the Enter key to return a newline (even on Windows), so consistently returning '\r' will avoid platform confusion.

Reply:
> the user may want to intercept and echo himself

At which point they could choose to do the same thing if they wanted to, no?

> I find it safe to return the '\r' to the user, since that is exactly how it works on Windows

To me that's like saying it's safe to have Environment.NewLine return "\r\n" on Unix because that's what it does on Windows.

Reply:
I think it's different. It's well known what newlines are for Windows and Unix.

If you'd ask a couple of users what character the enter key corresponds to on Windows or Unix, I expect you'd get a bunch of different answers.
I didn't figure out what the issue was since my expectation was the enter key corresponded to a \n character on Windows.
And on Linux, the enter key generates a \r character which is converted to a \n due to the ICRNL setting.

So I think it's a convention rather than something that will match a user expectation.
If we choose to return '\r' it gives us consistency and it makes existing code work.
I don't think we gain something by returning '\n' and get the opposite effects.

Reply:
Ok. I don't like it. But ok.